### DISCUSSION TITLE
locations.dat

### DISCUSSION THREAD

**Discussion by locatellidaniel** [THGSCST/Vices-and-Sins-MOD#2] [Discovery and Findings]
Hi,

Just wanted to share my findings about the ```locations.dat```

I've been using [ImHex](https://github.com/WerWolv/ImHex) a _great tool for this task_.

### Structure

So this is the structure, written in the ImHex pattern language.

```
#pragma pattern_limit 1000000

struct FileHeader {
	char unknown[0x3c] [[color("666666")]];
	le u16 group_count [[color("ffffff")]];
};

struct TileHeader {
	le u16 horizontal_frames;
	padding[2];
	le u16 vertical_frames;
	padding[2];
};

struct TileFrame {
	le u16 index;
	padding[2];
};

struct Tile {
	TileHeader header [[color("ff0000")]];
	TileFrame frames[header.horizontal_frames*header.vertical_frames - header.horizontal_frames/2] [[color("006600")]];	
};

struct TileGroup {
	Tile tiles[4];
	u8 group_id [[color("0000ff")]];
};

FileHeader fileheader @0x00;
TileGroup tile[fileheader.group_count] @ $;
```

### Tile 
So basically we have Tiles composed of frames (the 32x32 bitmaps already extracted from ```locations.tyl```).

The frame spatial positioning inside a tile is a bit tricky and it is better explained by a picture:

https://imgur.com/a/RnQkkdi

The width of this tile is actually 6 and the height is 5 yet the tile has only 27 frames defined.

This is because every even column (0,2,4...) does not have 5 frames but 4 (```height-1```) and an offset of 16px at the top. This explains the frame size count: ```header.horizontal_frames*header.vertical_frames - header.horizontal_frames/2```

### Tile groups

The Tiles are groupped with some kind of byte indicating sth. I've called this group_id but I don't know the real meaning.

### Sample Code

Since I'm not very fluent in C#, and I was just playing around, I have a JS script that extract all the tiles and save them to pngs.
Please let me know if you want me to share it somewhere.

I will now move to the other dat files and see if we can obtain more usable data so to extract or modify the people sprites for example.

**Comment by THGSCST** [comment]
That's a good find. Thanks for sharing!

Soon I will update the SPRTYL2PNG tool using the `locations.dat` information you provided. I'm very curious about the data near the end of this file, do you have any clues?

Don't worry about C# sample code, I still need to do some major refactoring in the tool code.

>I will now move to the other dat files and see if we can obtain more usable data so to extract or modify the people sprites for example.

I took a quick look at `PeopleSprites.dat`. It is possible to see chunks of sprites throughout the file, where the first four bytes give the sprite size (good example at `0x15871f`). The challenge is to understand how it organizes these chunks.
![image](https://user-images.githubusercontent.com/18038961/195755652-82f88d3b-6749-401b-a8c6-ba76918e5c48.png)

**Comment by locatellidaniel** [comment]
Hi!

> I'm very curious about the data near the end of this file, do you have any clues?

Not really, the bytes at the end as well as the header is still a mystery to me. I stopped looking after grabbing the tile data, yet there could still be sth there worth extracting.

> I took a quick look at PeopleSprites.dat. It is possible to see chunks of sprites throughout the file, where the first four bytes give the sprite size (good example at 0x15871f)

Yes! That's where the pixels starts, this is what I have so far:

```
#pragma pattern_limit 1000000

struct FileHeader {
	padding[0x3ec8a] [[hidden]]; // <<<< We need to figure this out
	le u16 sprite_chunk_count;
	padding[2] [[hidden]];
};

struct SpriteChunkHeader {
	le u16 width [[color("00ff00")]];
	le u16 height [[color("00dd00")]];
	padding[12] [[color("ffffff")]];   // <<<<  This also should be important
};

struct SpriteChunk {
	SpriteChunkHeader header;
	u8 pixel_data[header.width*header.height] [[color("666666")]];
};


FileHeader header @ 0x0;
SpriteChunk spritechunks[header.sprite_chunk_count] @ $;
```

PeopleSprites2, PeopleSprites3, and PeopleSprites4 (and also VehicleAndStreetFurnitureSprites) are slightly different so the offsets vary.
Yet I was able toe xtract enough data to render the 21880 sprites:

https://imgur.com/a/1i8AxPQ

Now the job should be to try to figure out how the rest of the data works, which I suspect makes everything fit together...

**Comment by locatellidaniel** [comment]
Hi @THGSCST 

I'm still working on it but I managed to find the *complete* structure of the PeopleSprites*.dat

Now i have to figure out how to put together the sprites. I didn't have enough time to think about it yet, but I'm pretty close.

Just wanted to let you know in case you're also working on it and want to share the progress.

**Reply by THGSCST** [reply]
Hello @locatellidaniel ,

Sorry for the late reply, just today I got some free time to update the SPRTYL2PNG tool, following your findings to implement tile group view and PNG export. Still missing the replace function, I'll be working on that soon.

About peoplesprites, I think that a new tool will have to be created, since it is necessary to animate many frames like walking and running, including the combinations of clothes and objects that the characters carry. I haven't done any debugging at the moment and I don't have any new information about the file format, if you can share details I would appreciate it.

Merry Christmas and Happy New Year!

**Reply by locatellidaniel** [reply]
Hi @THGSCST thanks for the greetings!

I checked your changes, it's looking great!

Replacing bitmap seems a hard task, but not as hard of what I'm seeing in PeopleSprites files :( 

I've opened a post with details of what I found as well as Imhex code and a html/js tool to try everything in the browser: https://github.com/THGSCST/Vices-and-Sins-MOD/discussions/3

Have a great New Year!
